### PR TITLE
SQL support for nested groupBys.

### PR DESCRIPTION
**Pull request by gianm**
Allows, for example, doing exact count distinct by writing:

```
SELECT COUNT(*) FROM (SELECT DISTINCT col FROM druid.foo)
```

Contrast with approximate count distinct, which is:

```
SELECT COUNT(DISTINCT col) FROM druid.foo
```

### CODE REVIEW

File: docs/content/querying/sql.md [hunk 1]
```diff
@@ -77,6 +77,20 @@ If `druid.sql.planner.useFallback` is enabled, full SQL is possible on metadata
 recommended in production since it can generate unscalable query plans. The JDBC driver allows accessing
 table and column metadata through `connection.getMetaData()` even if useFallback is off.
 
+### Approximate queries
+
+The following SQL queries and features may be executed using approximate algorithms:
+
+- `COUNT(DISTINCT col)` aggregations use [HyperLogLog](http://algo.inria.fr/flajolet/Publications/FlFuGaMe07.pdf), a
+fast approximate distinct counting algorithm. If you need exact distinct counts, you can instead use
+`SELECT COUNT(*) FROM (SELECT DISTINCT col FROM druid.foo)`, which will use a slower and more resource intensive exact
```
Comment (fjy) [inline docs/content/querying/sql.md]: 👍 on that suggestion

Comment (gianm) [inline docs/content/querying/sql.md]: That would be a nice feature, but it should be a different PR. Please file an issue if you want it :)

**Comment by fjy** [comment]
👍

File: docs/content/querying/sql.md (hunk 1 above)
Comment (jihoonson) [inline docs/content/querying/sql.md]: I also prefer this approach. Different behavior depending on query structure can make users confused.

Comment (gianm) [inline docs/content/querying/sql.md]: Fair enough, I agree that would be cool, but I don't think it makes sense to change DISTINCT aggs in this PR. All this PR is doing is adding the nested query feature, it's not making any changes to how DISTINCT aggs are handled.

Comment (jihoonson) [inline docs/content/querying/sql.md]: Ok. I'm reviewing the patch.

File: sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java [hunk 2]
```diff
@@ -95,64 +95,88 @@ public Void apply(final Row row)
...
+        throw new IllegalStateException("Can't use QueryDataSource without an outer groupBy query!");
+      }
+
+      executeGroupBy(queryBuilder, outerQuery, sink);
+      return;
+    }
+
+    final TimeseriesQuery timeseriesQuery = queryBuilder.toTimeseriesQuery(dataSource, sourceRowSignature);
+    if (timeseriesQuery != null) {
+      executeTimeseries(queryBuilder, timeseriesQuery, sink);
```
Comment (jihoonson) [inline sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java]: I think it would be better if we are able to know the query type ahead. But, I know this ```accumulate()``` method is just moved from ```DruidQueryBuilder``` with little changes, and adding query types will involve a lot of additional changes. Do you have any plan for this?

File: sql/src/main/java/io/druid/sql/calcite/rule/SelectRules.java [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+    @Override
+    public void onMatch(RelOptRuleCall call)
+    {
+      final Project project = call.rel(0);
+      final DruidRel druidRel = call.rel(1);
+
+      if (druidRel.getQueryBuilder().getSelectProjection() != null
+          || druidRel.getQueryBuilder().getGrouping() != null
+          || druidRel.getQueryBuilder().getLimitSpec() != null) {
+        return;
```
Comment (jihoonson) [inline sql/src/main/java/io/druid/sql/calcite/rule/SelectRules.java]: How about implementing ```public boolean matches(RelOptRuleCall call)```? I think this will be a better approach for avoiding not-matched rules early. 
I know these rule classes are just moved to here, but it will be good if we can improve them.

File: sql/src/main/java/io/druid/sql/calcite/rel/DruidQueryRel.java [hunk 4]
```diff
@@ -163,6 +188,20 @@ public RelWriter explainTerms(final RelWriter pw)
...
   {
-    return super.computeSelfCost(planner, mq).multiplyBy(0.1);
+    double cost = 1.0;
+
+    if (queryBuilder.getFilter() != null) {
+      cost *= 0.1;
+    }
+
+    if (queryBuilder.getGrouping() != null) {
+      cost *= 0.5;
```
Comment (jihoonson) [inline sql/src/main/java/io/druid/sql/calcite/rel/DruidQueryRel.java]: How about making these constants as static variables?

**Comment by gianm** [comment]
@jihoonson, on your test `testRecursivelyNestedGroupby`, that didn't work because it needs more merge buffers than the test allows, and since the merge buffer pool is blocking, the test blocks forever. I pushed a commit that adds your test, bumps up the test merge buffer pool to 3, adds a maxQueryCount config, and adds docs warning users about this potential deadlock in prod.

The doc blurb is:

> Note that groupBys require a separate merge buffer on the broker for each layer beyond the first layer of the groupBy. With the v2 groupBy strategy, this can potentially lead to deadlocks for groupBys nested beyond two layers, since the merge buffers are limited in number and are acquired one-by-one and not as a complete set. At this time we recommend that you avoid deeply-nested groupBys with the v2 strategy. Doubly-nested groupBys (groupBy -> groupBy -> table) are safe and do not suffer from this issue. If you like, you can forbid deeper nesting by setting `druid.sql.planner.maxQueryCount = 2`.

File: sql/src/main/java/io/druid/sql/calcite/rule/SelectRules.java (hunk 3 above)
Comment (gianm) [inline sql/src/main/java/io/druid/sql/calcite/rule/SelectRules.java]: Sounds good, I'll do that. I didn't think of this before.

File: sql/src/main/java/io/druid/sql/calcite/rel/DruidQueryRel.java (hunk 4 above)
Comment (gianm) [inline sql/src/main/java/io/druid/sql/calcite/rel/DruidQueryRel.java]: Sure thing.

File: sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java (hunk 2 above)
Comment (gianm) [inline sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java]: I thought of having just one `toQuery` method, but the problem with that is then when we want to execute the query, we need to pair it with the correct execution strategy for that query (select needs to issue multiple queries for pagination, all query types have different result formats, etc). So that's why `accumulate` checks each possible query type individually.

I don't have a plan for changing this but I am open to change. Maybe DruidQueryBuilder should return both a Query and some kind of execution strategy object for it?

**Comment by gianm** [comment]
@jihoonson, I just pushed commits for the rest of your comments, please let me know what you think. thanks for the review.

**Comment by gianm** [comment]
Raised #3819 for the deeply nested groupBy thing.

File: sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java (hunk 2 above)
Comment (jihoonson) [inline sql/src/main/java/io/druid/sql/calcite/rel/QueryMaker.java]: Ok. We can change later if it needs.

**Comment by jihoonson** [comment]
Thanks! The latest patch looks good to me.

**Comment by gianm** [comment]
Resolved conflicts.

**Comment by jon-wei** [comment]
:+1:
